plugins/contiv/ipam/ipam.go

```diff
@@ -243,29 +246,50 @@ func (i *IPAM) NextPodIP(podID string) (net.IP, error) {
 		return nil, err
 	}
 
+	last := i.lastAssigned + 1
 	// iterate over all possible IP addresses for pod network prefix
-	// and take first not assigned IP
+	// start from the last assigned and take first available IP
 	prefixBits, totalBits := i.podNetworkIPPrefix.Mask.Size()
 	maxSeqID := 1 << uint(totalBits-prefixBits) //max IP addresses in network range
-	for j := 1; j < maxSeqID; j++ {             // zero ending IP is reserved for network => skip seqID=0
-		if j == podGatewaySeqID {
-			continue // gateway IP address can't be assigned as pod
-		}
-		if _, found := i.assignedPodIPs[networkPrefix+uint32(j)]; found {
-			continue // ignore already assigned IP addresses
+	for j := last; j < maxSeqID; j++ {          // zero ending IP is reserved for network => skip seqID=0
+		ipForAssign, success := i.tryToAllocatePodIP(j, networkPrefix, podID)
+		if success {
+			i.lastAssigned = j
+			return ipForAssign, nil
 		}
-		i.assignedPodIPs[networkPrefix+uint32(j)] = podID
-		//TODO set etcd for new assigned value
+	}
 
-		ipForAssign := uint32ToIpv4(networkPrefix + uint32(j))
-		i.logger.Infof("Assigned new pod IP %s", ipForAssign)
-		i.logAssignedPodIPPool()
-		return ipForAssign, nil
+	// iterate from the range start until lastAssigned
+	for j := 1; j < last; j++ { // zero ending IP is reserved for network => skip seqID=0
+		ipForAssign, success := i.tryToAllocatePodIP(j, networkPrefix, podID)
+		if success {
+			i.lastAssigned = j
+			return ipForAssign, nil
+		}
 	}
 
 	return nil, fmt.Errorf("No IP address is free for assignment. All IP addresses for pod network %v are already assigned", i.podNetworkIPPrefix)
 }
 
+// tryToAllocatePodIP checks whether the IP at the given index is available.
+func (i *IPAM) tryToAllocatePodIP(index int, networkPrefix uint32, podID string) (assignedIP net.IP, success bool) {
+	if index == podGatewaySeqID {
+		return nil, false // gateway IP address can't be assigned as pod
+	}
+	ip := networkPrefix + uint32(index)
+	if _, found := i.assignedPodIPs[ip]; found {
```

Comment:
i don't see `assignedPodIPs` is initiated from etcd, how you make sure it's not being used after restarting ?